infrastructure/resource_group_level/modules/storage.bicep

```diff
@@ -0,0 +1,421 @@
+// param virtualNetworks_vnet_ukef_uks_prod_externalid string = '/subscriptions/08887298-3821-49f0-8303-f88859c12b9b/resourceGroups/rg-ukef-uks-network/providers/Microsoft.Network/virtualNetworks/vnet-ukef-uks-prod'
+
+param appServicePlanEgressSubnetId string
+param gatewaySubnetId string
+param privateEndpointsSubnetId string
+
+@description('IPs or CIDRs still allowed to access the storage if the default action is Deny')
+param allowedIps array
+
+@description('Is public access to the storage account allowed or denied for evertone')
+@allowed(['Allow', 'Deny'])
+param networkAccessDefaultAction string
+
+var storageAccountName = 'tfs${environment}storage'
+var ipRules = [for ip in allowedIps: {
+  value: ip
+  action: 'Allow'
+}]
+
+resource storageAccount 'Microsoft.Storage/storageAccounts@2022-09-01' = {
+  name: storageAccountName
+  location: location
+  tags: {
+    Environment: 'Preproduction'
+  }
+  sku: {
+    name: 'Standard_ZRS'
+  }
+  kind: 'StorageV2'
+  properties: {
+    defaultToOAuthAuthentication: false
+    publicNetworkAccess: 'Enabled'
+    minimumTlsVersion: 'TLS1_2'
+    allowBlobPublicAccess: false
+    allowSharedKeyAccess: true
+    networkAcls: {
+      resourceAccessRules: []
+      bypass: 'AzureServices'
+      virtualNetworkRules: [
+      // TODO:FN-417 why is there a production subnet here? Only include if proved necessary
+        // {
+        //   id: '${virtualNetworks_vnet_ukef_uks_prod_externalid}/subnets/snet-wvd'
+        //   action: 'Allow'
+        // }
+        {
+          id: appServicePlanEgressSubnetId
+          action: 'Allow'
+        }
+        {
+          id: gatewaySubnetId
+          action: 'Allow'
+        }
+        {
+          id: privateEndpointsSubnetId
+          action: 'Allow'
+        }
+      ]
+      ipRules: ipRules
+      defaultAction: networkAccessDefaultAction
+    }
+    supportsHttpsTrafficOnly: true
+    encryption: {
+      services: {
+        file: {
+          keyType: 'Account'
+          enabled: true
+        }
+        blob: {
+          keyType: 'Account'
+          enabled: true
+        }
+      }
+      keySource: 'Microsoft.Storage'
+    }
+    accessTier: 'Hot'
+  }
+}
+
+resource defaultBlobService 'Microsoft.Storage/storageAccounts/blobServices@2022-09-01' = {
+  parent: storageAccount
+  name: 'default'
+  properties: {
+    containerDeleteRetentionPolicy: {
+      enabled: true
+      days: 3
+    }
+    cors: {
+      corsRules: []
+    }
+    deleteRetentionPolicy: {
+      allowPermanentDelete: false
+      enabled: true
+      days: 3
+    }
+  }
+}
+
+resource defaultFileService 'Microsoft.Storage/storageAccounts/fileServices@2022-09-01' = {
+  parent: storageAccount
+  name: 'default'
+  properties: {
+    cors: {
+      corsRules: []
+    }
+  }
+}
+
+resource queueService 'Microsoft.Storage/storageAccounts/queueServices@2022-09-01' = {
+  parent: storageAccount
+  name: 'default'
+  properties: {
```

Comment:
It's probably best to remove empty initialisations as they just add clutter. I'll do so if it behaves correctly.

Reply:
sounds good